flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestClientFactory.java

```diff
@@ -131,6 +138,43 @@ void destroyPartitionRequestClient(ConnectionID connectionId, PartitionRequestCl
 		clients.remove(connectionId, client);
 	}
 
+	private void connectChannelWithRetry(ConnectingChannel connectingChannel,
+										 ConnectionID connectionId, boolean needConnect)
+		throws IOException, InterruptedException {
+		int count = 0;
+		Exception exception = null;
+		do {
+			try {
+				if (needConnect) {
+					LOG.info("Connecting to {} at {} attempt", connectionId.getAddress(), count);
+					nettyClient.connect(connectionId.getAddress()).addListener(connectingChannel);
+				}
+
+				NettyPartitionRequestClient client = connectingChannel.waitForChannel();
+
+				clients.replace(connectionId, connectingChannel, client);
+				return;
+			} catch (IOException | ChannelException e) {
+				LOG.error("Failed {} times to connect to {}", count, connectionId.getAddress(), e);
+				ConnectingChannel newConnectingChannel = new ConnectingChannel(connectionId, this);
+				clients.replace(connectionId, connectingChannel, newConnectingChannel);
+				Object old = clients.get(connectionId);
+				if (old instanceof ConnectingChannel) {
+					connectingChannel = (ConnectingChannel) old;
+				} else {
+					return;
+				}
+
+				exception = e;
+			}
+			count++;
+		} while (count <= retryNumber);
```

Comment:
The interaction here between multiple requesters is getting really complicated. Especially this part in the catch block seems strange. I think it might be working, but I'm not sure. It's quite hard to reason about and it's missing test coverage.

On the other hand following @zhijiangW 's suggestion with using plain `HashMap` with manually synchronisation and just blocking subsequent threads from entering until reconnection loop completes sounds much simpler, but there is that issue of how to implement other methods of this class.

All in all, I think I would be +1 for going as it is, given that we would have also another test case, similar to the existing `testNettyClientConnectRetry`, but which spawns couple of threads trying to do the same thing:
```
factory.createPartitionRequestClient(serverAddress);
```
?

Reply:
Add a multiple thread test case for client creation.